router.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package ringpop
+
+import (
+	"sync"
+
+	"github.com/uber/ringpop-go/swim"
+)
+
+// Replication constants
+type ReplicationFactor int
+
+const (
+	ALL ReplicationFactor = -1 * (iota + 1)
+	// TODO implement quorum
+	// QUORUM
+)
+
+type router struct {
+	ringpop RingpopInterface
+	factory ClientFactory
+
+	rw          sync.RWMutex
+	clientCache map[string]interface{}
+}
+
+type Router interface {
+	GetClients(key string, n ReplicationFactor, fanout ReplicationFactor) []interface{}
+}
+
+type ClientFactory interface {
+	MakeClient(hostport string) interface{}
+}
+
+func NewRouter(rp RingpopInterface, f ClientFactory) Router {
+	r := &router{
+		ringpop:     rp,
+		factory:     f,
+		clientCache: make(map[string]interface{}),
+	}
+	rp.RegisterListener(r)
+	return r
+}
+
+func (r *router) HandleEvent(event interface{}) {
+	switch event := event.(type) {
+	case swim.MemberlistChangesReceivedEvent:
+		for _, change := range event.Changes {
+			r.handleChange(change)
+		}
+	}
+}
+
+func (r *router) handleChange(change swim.Change) {
+	switch change.Status {
+	case swim.Faulty, swim.Leave:
+		r.removeClient(change.Address)
+	}
+}
+
+// Find the destinations for a certain key with the given N and Fanout
+func (r *router) getDestinations(key string, n ReplicationFactor, fanout ReplicationFactor) []string {
```

Comment:
These replication controls were slightly confusing to me at first and that's because I thought you were going for Riak-type CAP controls. The way the N and R/W values work in Riak is as follows (from Riak docs: http://docs.basho.com/riak/1.2.1/tutorials/fast-track/Tunable-CAP-Controls-in-Riak/#A-Primer-on-N-R-and-W):

N-Value
```
All data stored in Riak will be replicated to a number of nodes in the cluster according to the N value (n_val) property set on the bucket. By default, Riak chooses an n_val of “3” for you. This means that data stored in the bucket will be replicated to three different nodes, thus storing three copies. For this to be effective, you need at least three physical nodes in your cluster. (We can, however, demonstrate its merits with local nodes.)
```

R-Value and Read Failure Tolerance
```
Riak allows the client to supply an “R value” on each direct fetch. The R value represents the number of Riak nodes which must return results for a read before the read is considered successful. This allows Riak to provide read availability even when nodes are down or laggy.
```

This is not the way you've defined these replication controls. And it also goes against the grain of how we've previously used N, R and W values in Ringpop Replicator. We _should_ get closer to Riak's definition. In the current implementation, `fanout` does not actually control the fanout factor.